### PR TITLE
Docs/multi extauth

### PR DESCRIPTION
**Pull request by bcollard**
# Description

Docs only

Explain when to use namedExtAuth vs passThrough

BOT NOTES: 
resolves https://github.com/solo-io/gloo/pull/5733

### CODE REVIEW

**Comment by bcollard** [comment]
it's ready to review. Expect some delay on my side, I'll be on vacation for the next few days.

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+description: Configure multiple External Authorization servers. Decide which one to use at the route level.
+---
+
+Gloo Edge Enterprise comes with an External Authorization service, with full-featured security policies, like OpenID Connect, Open Policy Agent, OAuth 2 scopes and a few more. Also, using the so-called **passthrough** system, you can extend the authorization logic and call your own authentication or authorization service.
+
+There are certain cases where it's more relevant to directly plug the `ext_authz` Envoy filter to your own authorization service. This article describes how to configure such connections between the Envoy data-plane and third-party authorization services.
+
+## Calling an external authorization service
+
+Say you are running an corporate access management service, that is in charge of returning new claims, like the `role` of the user who is authenticated.
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:13]:
```suggestion
Say you are running a corporate access management service that is in charge of returning new claims, like the `role` of the user who is authenticated.
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+With Gloo Edge Enterprise, there are two - non-exclusive - ways of calling such an _external_ service:
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:14]:
```suggestion
With Gloo Edge Enterprise, you can call the external service in two, non-exclusive ways, as shown in the following figure.
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+- Option A: use the **Passthrough Auth** [plugin](http://localhost:1313/guides/security/auth/extauth/passthrough_auth/), which comes with all the ExtAuth options (see `AuthConfig > passthrough`)
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:15]:
```suggestion
- Option A: use the **Passthrough Auth** [plug-in]({{< versioned_link_path fromRoot="/guides/security/auth/extauth/passthrough_auth/" >}}), which comes with all the ExtAuth options (see `AuthConfig > passthrough`)
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+- Option B: define an additional ExtAuth server, using `namedExtAuth` and call it on the routes of your choice
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:16]:
```suggestion
- Option B: define an additional ExtAuth server with the `namedExtAuth` setting, and call the additional ExtAuth server on the routes of your choice
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+![Calling an external authorization service](./two-options-external-authz-service.png)
+
+While there is an extra network hop in the first case, you will be able to leverage the `AuthConfig` CR, and also to easily pass information between the authN/Z steps. See the [dedicated guide]({{% versioned_link_path fromRoot="/guides/security/auth/extauth/passthrough_auth/grpc/" %}}) for more information.
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:20]:
```suggestion
In the first option, the connection has an extra network hop, but you can use the `AuthConfig` CR and more easily pass information between the authentication and authorization steps. For more information, see the [gRPC Passthrough Auth guide]({{% versioned_link_path fromRoot="/guides/security/auth/extauth/passthrough_auth/grpc/" %}}).
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+### Option A - Using the passthrough system
+
+With Option A, you can compose your authentication and authorization workflow with the `AuthConfig` Custom Resource.
+For instance, you can have a block that leverages the builtin OpenID Connect plugin and another block where you define how to call your external authorization service.
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:25]:
```suggestion
For instance, you can have a block that leverages the built-in OpenID Connect plug-in and another block where you define how to call your external authorization service. 
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+![Calling an external authorization service](./two-options-external-authz-service.png)
+
+While there is an extra network hop in the first case, you will be able to leverage the `AuthConfig` CR, and also to easily pass information between the authN/Z steps. See the [dedicated guide]({{% versioned_link_path fromRoot="/guides/security/auth/extauth/passthrough_auth/grpc/" %}}) for more information.
+
+### Option A - Using the passthrough system
+
+With Option A, you can compose your authentication and authorization workflow with the `AuthConfig` Custom Resource.
+For instance, you can have a block that leverages the builtin OpenID Connect plugin and another block where you define how to call your external authorization service. 
+
+Below is an common use case where Gloo achieves the following steps:
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:27]:
```suggestion
The following diagram depicts a common use case wherein Gloo:
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+- use OIDC to authenticate the end-user
+- use the passthrough system to populate the request with an additional JWT, having authorization-centered claims (roles, etc.)
+- use the JWT policy to verify the signature of the newly created JWT
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:30]:
```suggestion
- uses OIDC to authenticate the end user
- uses the passthrough system to populate the request with an additional JWT, having authorization-centered claims (roles, etc.)
- uses the JWT policy to verify the signature of the newly created JWT
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+![Compose your AuthN + AuthZ security workflow](./authconfig-oidc-and-passthrough.png)
+
+Also, note you can pass values between the `AuthConfig` blocks using the `State` map. More info in this [section]({{% versioned_link_path fromRoot="/guides/security/auth/extauth/passthrough_auth/grpc/#sharing-state-with-other-auth-steps" %}}).
+
+You can use either the [gRPC Passthrough Auth]({{% versioned_link_path fromRoot="/guides/security/auth/extauth/passthrough_auth/grpc/" %}}) protocol or [HTTP Passthrough Auth]({{% versioned_link_path fromRoot="/guides/security/auth/extauth/passthrough_auth/http/" %}}).
+
+### Option B - Using namedExtAuth
+
+With this new option (available from version 1.9), you can define additional ExtAuthZ servers in the Gloo Edge settings.
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:40]:
```suggestion
In Gloo Edge Enterprise version 1.9 or later, you can define additional ExtAuthZ servers in the Gloo Edge settings.
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+For that, you must register your authorization servers in the "default" `Settings` custom resource. See the [API reference](https://docs.solo.io/gloo-edge/latest/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/settings.proto.sk/#settings).
+
+Below is an example:
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:43]:
```suggestion
For that, you must register your authorization servers in the "default" `Settings` custom resource, as shown in the following example. For more information, see the [API reference](https://docs.solo.io/gloo-edge/latest/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/settings.proto.sk/#settings).

```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+spec:
+  extauth:
+    extauthzServerRef:
+      name: extauth
+      namespace: gloo-system
+    transportApiVersion: V3
+    userIdHeader: x-user-id
+  namedExtauth:
+    customHttpAuthz: # custom name
+      extauthzServerRef: # define where is running the third-party authorization server
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:56]:
```suggestion
      extauthzServerRef: # define the third-party authorization server runs in your cluster
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+With Option A, you can compose your authentication and authorization workflow with the `AuthConfig` Custom Resource.
+For instance, you can have a block that leverages the builtin OpenID Connect plugin and another block where you define how to call your external authorization service. 
+
+Below is an common use case where Gloo achieves the following steps:
+- use OIDC to authenticate the end-user
+- use the passthrough system to populate the request with an additional JWT, having authorization-centered claims (roles, etc.)
+- use the JWT policy to verify the signature of the newly created JWT
+
+![Compose your AuthN + AuthZ security workflow](./authconfig-oidc-and-passthrough.png)
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:32]:
```suggestion
<figure><img src="{{% versioned_link_path fromRoot="/guides/security/auth/multi_authz/authconfig-oidc-and-passthrough.png" %}}">
<figcaption style="text-align:center;font-style:italic">Figure: Configure your authentication and authorization security workflow</figcaption></figure>
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+            namespace: gloo-system
+      options:
+        extauth:
+          customAuth: # set the authorization system at the route level
+            name: customHttpAuthz # one of the `namedExtAuth` 
+{{< /highlight >}}
+
+## The gRPC specification
+
+Be it with the **Passthrough Auth** option or with the **namedExtAuth** option, you must conform to the Envoy specification for the external Authorization service: https://github.com/envoyproxy/envoy/blob/main/api/envoy/service/auth/v3/external_auth.proto
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:98]:
```suggestion
For both the **Passthrough Auth** option and the **namedExtAuth** option, you must conform to the Envoy specification for the external Authorization service. For more information, see the following resources.
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
+You will find more details in the [dedicated article]({{% versioned_link_path fromRoot="/guides/security/auth/extauth/passthrough_auth/grpc/" %}}). Also, there are examples in this GitHub repository: https://github.com/solo-io/gloo/tree/master/docs/examples/grpc-passthrough-auth/pkg/auth/v3
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:100]:
```suggestion
* [Envoy external auth proto](https://github.com/envoyproxy/envoy/blob/main/api/envoy/service/auth/v3/external_auth.proto)
* [gRPC Passthrough Auth guide]({{% versioned_link_path fromRoot="/guides/security/auth/extauth/passthrough_auth/grpc/" %}})
* [GitHub repo with Go examples](https://github.com/solo-io/gloo/tree/master/docs/examples/grpc-passthrough-auth/pkg/auth/v3)
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+
+## Managing headers
+
+If you want to add new headers to the request from your external authorization service, this is possible.
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:104]:
```suggestion
You can add new headers to the request from your external authorization service.
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+
+### gRPC mode
+
+In the case of gRPC, you will rely on the [protobuf specification](https://github.com/envoyproxy/envoy/blob/main/api/envoy/service/auth/v3/external_auth.proto#L76) and you need [to populate]({{% versioned_link_path fromRoot="/guides/security/auth/extauth/passthrough_auth/grpc/#sharing-state-with-other-auth-steps" %}}) the `OkHttpResponse` with the new headers. They will be visible to the rest of the filter chain and they will be passed to the upstream service. See [this section]({{% versioned_link_path fromRoot="/guides/security/auth/extauth/passthrough_auth/grpc/#sharing-state-with-other-auth-steps" %}}) of the documentation.
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:108]:
```suggestion
For gRPC, use the [protobuf specification](https://github.com/envoyproxy/envoy/blob/main/api/envoy/service/auth/v3/external_auth.proto#L76) and [populate the `OkHttpResponse`]({{% versioned_link_path fromRoot="/guides/security/auth/extauth/passthrough_auth/grpc/#sharing-state-with-other-auth-steps" %}}) with the new headers. The new headers are visible to the rest of the filter chain and are passed to the upstream service. For more information, see [Sharing state with other auth steps]({{% versioned_link_path fromRoot="/guides/security/auth/extauth/passthrough_auth/grpc/#sharing-state-with-other-auth-steps" %}}).
```

File: docs/content/guides/security/auth/multi_authz/_index.md [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+
+### HTTP mode
+
+With both the **HTTP passthrough** option and the **namedExtAuth** with HTTP option, if you want to send new headers to the rest of the filter chain or to the upstream service, then just add them to the authorization response and they will be merged into the original request before being forwarded upstream.
```
Comment (artberger) [inline docs/content/guides/security/auth/multi_authz/_index.md:112]:
```suggestion
With both the **HTTP passthrough** option and the **namedExtAuth** with HTTP option, if you want to send new headers to the rest of the filter chain or to the upstream service, then just add the headers to the authorization response. Then, the headers are merged into the original request before being forwarded upstream.
```

**Review by artberger — COMMENTED** [review]
Thanks for adding this doc! I left some suggestions for grammar and style.

File: docs/content/guides/security/auth/multi_authz/_index.md (hunk 11 above)
Comment (bcollard) [reply docs/content/guides/security/auth/multi_authz/_index.md:56]: technically, the authZ server can run everywhere

**Review by artberger — APPROVED** [review]

**Pull request merged**